### PR TITLE
LUCENE-10260: Luke's about window no longer shows version number

### PR DESCRIPTION
**Pull request by dweiss**

### CODE REVIEW

**Comment by uschindler** [comment]
Looks fine, although I think the class LucenePackage should possibly go away completely in 9.

We can add a getImplementationVersion to our standard Version class which looks like a better place.

**Comment by dweiss** [comment]
Sure, nuke or move anything you want - I just wanted to show that this works.

File: lucene/core/src/java/org/apache/lucene/LucenePackage.java [hunk 1]
```diff
@@ -16,13 +16,53 @@
...
+
+      Package p = LucenePackage.class.getPackage();
+      version = p.getImplementationVersion();
+
+      if (version == null) {
+        var module = LucenePackage.class.getModule();
+        if (module.isNamed()) {
+          // Running as a module? Try parsing the manifest manually.
+          InputStream is = null;
+          try {
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/LucenePackage.java:46]: Better use try with resources

**Review by uschindler — CHANGES_REQUESTED** [review]

File: lucene/core/src/java/org/apache/lucene/LucenePackage.java (hunk 1 above)
Comment (dweiss) [reply lucene/core/src/java/org/apache/lucene/LucenePackage.java:46]: NPE?

Comment (dweiss) [reply lucene/core/src/java/org/apache/lucene/LucenePackage.java:46]: (if manifest is not there for some reason).

Comment (uschindler) [reply lucene/core/src/java/org/apache/lucene/LucenePackage.java:46]: Try with resources works with null.

Comment (uschindler) [reply lucene/core/src/java/org/apache/lucene/LucenePackage.java:46]: https://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-14.20.3

> A resource is closed only if it initialized to a non-null value.

Comment (dweiss) [reply lucene/core/src/java/org/apache/lucene/LucenePackage.java:46]: Ah! You learn every day, thanks!

**Comment by dweiss** [comment]
> We can add a static getImplementationVersion to our standard Version class which looks like a better place.

Done. I verified that both of these work and show the implementation version:

{code}
java -jar lucene/luke/build/lucene-luke-10.0.0-SNAPSHOT/lucene-luke-10.0.0-SNAPSHOT-standalone.jar
java --module-path lucene/luke/build/lucene-luke-10.0.0-SNAPSHOT --add-modules org.apache.logging.log4j -m lucene.luke
{code}

**Comment by dweiss** [comment]
This is what is shows if started from the IDE (when no module and no manifest is preset). I think it's fine:
image

**Review by uschindler — APPROVED** [review]

**Comment by uschindler** [comment]
Hi,

this looks good to me. I am a bit hesitant about Version.getImplementationVersion(), as the Version class also has a LATEST and CURRENT_VERSION symbol. Mabye rename Version.getPackageImplementationVersion().

But I am happy of the functionality, but because of this as a new API addition, we should think about naming.

File: lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java [hunk 2]
```diff
@@ -172,16 +172,15 @@ private JPanel footer() {
...
   }
 
-  private static final String LUCENE_IMPLEMENTATION_VERSION =
-      LucenePackage.get().getImplementationVersion();
+  private static final String LUCENE_IMPLEMENTATION_VERSION = Version.getImplementationVersion();
 
   private static final String LICENSE_NOTICE =
       "<p>[Implementation Version]</p>"
           + "<p>"
           + (Objects.nonNull(LUCENE_IMPLEMENTATION_VERSION) ? LUCENE_IMPLEMENTATION_VERSION : "")
```
Comment (uschindler) [inline lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java:180]: I think the nullcheck is not needed anymore.

**Comment by dweiss** [comment]
The more I think about it, the more I'm wondering why this method is needed... Luke receives the same metadata (manifest) as the rest of Lucene - we could just display this instead (from our own module), remove the extra method from the core and, just in case, display Version.VERSION_CURRENT (or whatever it's called) to indicate which Lucene JAR Luke is working against. I'm not even sure a scenario of running Luke against a different Lucene version is supported, but just in case.

**Comment by uschindler** [comment]
Hi Dawid,
i think the important thing was in former days of Luke to see the implementation version of lucene-core.jar. Also other projects like Elasticsearch or Solr want to show this. So I think the method in Version is fine.
We should at least add a cahenges entry that we removed the LucenePackage class and added a implementation version getter in lucene's core.
About luke: Of course luke could only show its version, but for the reasons above i tend to think that we should provide a way for external projects to show the lucene version they are linking to.
I think code is same, but having a Version.getImplementationVersion() is a good idea to have. We provided it for years (since Lucene v1.0).

**Comment by dweiss** [comment]
ok, sure. please go ahead and commit those minor changes/ merge them in if you like - I'll be home late.

File: lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java [hunk 3]
```diff
@@ -173,15 +172,15 @@ private JPanel footer() {
   }
 
   private static final String LUCENE_IMPLEMENTATION_VERSION =
-      LucenePackage.get().getImplementationVersion();
+      Version.getPackageImplementationVersion();
```
Comment (mocobeta) [inline lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java:175]: (Question) If Luke runs on module mode and `MANIFEST.MF` does not exist in lucene-core jar for some reason (on development?), Luke fails to start...?

File: lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java [hunk 4]
```diff
@@ -173,15 +172,15 @@ private JPanel footer() {
...
 
   private static final String LICENSE_NOTICE =
       "<p>[Implementation Version]</p>"
```
Comment (mocobeta) [inline lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java:178]: Browsing through the conversation here:
Just to be more specific, how about changing this to `Lucene Implementation Version` if we see Luke is a standalone application strictly separated from the Lucene core module?

**Comment by mocobeta** [comment]
> I'm not even sure a scenario of running Luke against a different Lucene version is supported, but just in case.

(Yes there are some needs like that - some people want to browse an old Lucene codec with the latest Luke - though I think supporting multiple Lucene API versions is not feasible.)

File: lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java (hunk 3 above)
Comment (dweiss) [reply lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java:175]: Version.getPackageImplementationVersion will always return a non-null string - an "unknown" string is returned if manifest is not present.

File: lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java (hunk 4 above)
Comment (dweiss) [reply lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java:178]: I think it makes sense.

Comment (dweiss) [reply lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java]: Done.

Comment (mocobeta) [reply lucene/luke/src/java/org/apache/lucene/luke/app/desktop/components/dialog/menubar/AboutDialogFactory.java]: Thanks!

**Pull request merged**